src/Bicep.Core/SemanticModel/SymbolValidator.cs

```diff
@@ -0,0 +1,100 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Bicep.Core.Diagnostics;
+using Bicep.Core.Parser;
+using Bicep.Core.Syntax;
+using Bicep.Core.Text;
+using Bicep.Core.TypeSystem;
+
+namespace Bicep.Core.SemanticModel
+{
+    public class SymbolValidator
+    {
+        private delegate IEnumerable<string> GetNameSuggestions();
+        private delegate ErrorDiagnostic GetMissingNameError(DiagnosticBuilder.DiagnosticBuilderInternal builder, string? suggestedName);
+
+        public static Symbol ValidateNamespaceQualifiedFunction(FunctionFlags allowedFlags, Symbol? foundSymbol, IdentifierSyntax identifierSyntax, NamespaceSymbol namespaceSymbol)
```

Comment:
nit - `Validate*` implies return type `void` to me for some reason. Can we call all of these methods `Resolve*` or something similar?